contrib/repro_test.bzl

```diff
@@ -0,0 +1,243 @@
+A test image is considered to be reproducible when both instances of the same
+image have the same digest and ID. On a mismatch of one of those values, the
+test fails and the container_diff tool
+(https://github.com/GoogleContainerTools/container-diff#container-diff)
+produces the summary of their differences.
+
+Args:
+    test_image: A container_image target relative to the project's root to test
+                for determinism.
+    src_project_files: A target that exposes the files of the Bazel project to
+                       build the test_image in (e.g. filegroup target).
+    base: An image target to build and reproduce the test_image in.
+    container_diff_args: (optional) Args to the container_diff tool as specified here:
+                         https://github.com/GoogleContainerTools/container-diff#quickstart
+
+Example:
+
+container_repro_test(
+    name = "set_cmd_repro_test",
+    # Note that this target must always be relative to the project's root.
+    test_image = "//tests/container:set_cmd",
+    base = "@bazel_latest//image",
+    src_project_files = "//:src_project",
+    container_diff_args: [
+        "--type=history",
+        "--type=file",
+        "--type=size",
+        "--type=rpm",
+        "--type=apt",
+        "--type=pip",
+        "--type=node",
+    ],
+)
+"""
+
+load("@base_images_docker//util:run.bzl", _extract = "extract")
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("//container:container.bzl", _container = "container")
+load("//skylib:filetype.bzl", container_filetype = "container")
+
+def _impl(ctx):
+    """Core implementation of container_repro_test"""
+
+    # Check for existence of the WORKSPACE file in the files provided.
+    src_project_files = ctx.files.src_project_files
+    workspace_file = None
+    for f in src_project_files:
+        if f.basename == "WORKSPACE":
+            workspace_file = f
+            break
+
+    if not workspace_file:
+        fail("A WORKSPACE file must be among the src_project_files provided files.")
+
+    name = ctx.attr.name
+
+    # Record the absolute path to the provided WORKSPACE file and assume it to
+    # be the root of the Bazel project to build the test_image in.
+    proj_root = ctx.actions.declare_file(name + "_src_project_root.txt")
+    ctx.actions.run_shell(
+        command = "readlink -f %s | xargs dirname >> %s" % (workspace_file.path, proj_root.path),
+        inputs = [workspace_file],
+        outputs = [proj_root],
+    )
+
+    src_paths = [f.path for f in src_project_files]
+    proj_tar = ctx.actions.declare_file(name + "_src_project.tar")
+
+    # Tar the provided source project files to mount into the image that will
+    # be used to build and repro the test_image in.
+    ctx.actions.run_shell(
+        command = "readlink -f %s | xargs tar -cvf %s" % (" ".join(src_paths), proj_tar.path),
+        inputs = src_project_files,
+        outputs = [proj_tar],
+    )
+
+    image_tar = ctx.actions.declare_file(name + ".tar")
```

Comment:
we probably need to clean up this very large file at the end of the execution of the rule?

Reply:
The only way I could think of doing this is using `ctx.actions.run_shell`. But it neither produces an error nor deletes the file when using the `rm` command.

Reply:
ok, for now lets just leave a todo to try to remove it later.

Reply:
Done